src/ServiceBasedMerchant/ServiceBasedMerchantState.php

```diff
@@ -0,0 +1,240 @@
+		}
+
+		$this->clear_physical_products_cache();
+		// Delete the option so it gets recalculated on next check, or recalculate now
+		$this->options->delete( OptionsInterface::IS_SERVICE_BASED_MERCHANT );
+		$this->calculate_service_based_merchant();
+	}
+
+	/**
+	 * Clear cache when a post is changed, but only if it's a product.
+	 *
+	 * @param int    $post_id         Post ID.
+	 * @param string $previous_status Optional previous status (passed by wp_trash_post/untrashed_post hooks).
+	 */
+	public function maybe_clear_cache_on_post_change( int $post_id, $previous_status = null ): void {
+		$post_type = get_post_type( $post_id );
+		if ( 'product' === $post_type ) {
+			$this->clear_physical_products_cache();
+			// Delete the option so it gets recalculated on next check, or recalculate now
+			$this->options->delete( OptionsInterface::IS_SERVICE_BASED_MERCHANT );
+			$this->calculate_service_based_merchant();
+		}
+	}
+
+	/**
+	 * Check if the store is service-based.
+	 *
+	 * If the option has not been calculated yet, it will be calculated automatically
+	 * based on whether the store has physical products.
+	 *
+	 * @return bool True if service-based, false otherwise.
+	 */
+	public function is_service_based_merchant(): bool {
+		$option_value = $this->options->get( OptionsInterface::IS_SERVICE_BASED_MERCHANT );
+
+		// If option is null, calculate it now.
+		if ( null === $option_value ) {
+			return $this->calculate_service_based_merchant();
+		}
+
+		return (bool) $option_value;
+	}
+
+	/**
+	 * Calculate and save whether the store is service-based.
+	 *
+	 * This method should be called early in the onboarding process to determine
+	 * if the store has physical products that require shipping.
+	 *
+	 * @return bool True if service-based, false otherwise.
+	 */
+	public function calculate_service_based_merchant(): bool {
+		// Clear the cache before calculating to ensure fresh data.
+		$this->clear_physical_products_cache();
+
+		$service_based = ! $this->has_physical_products();
+
+		$this->options->update( OptionsInterface::IS_SERVICE_BASED_MERCHANT, $service_based );
+		return $service_based;
+	}
+
+	/**
+	 * Check if the store has at least one physical product that requires shipping.
+	 *
+	 * Results are cached to improve performance. The cache can be bypassed or modified
+	 * using the 'woocommerce_gla_has_physical_products' filter.
+	 *
+	 * @return bool True if at least one physical product requiring shipping is found, false otherwise.
+	 */
+	public function has_physical_products(): bool {
+		/**
+		 * Filter to bypass or modify the cached result of has_physical_products.
+		 *
+		 * @since 1.0.0
+		 *
+		 * @param bool|null $cached_result The cached result, or null to use default caching behavior.
+		 * @return bool|null True if physical products exist, false otherwise, or null to use cache.
+		 */
+		$filtered_result = apply_filters( 'woocommerce_gla_has_physical_products', null );
```

Comment:
@jamesmorrison I don't think theres a need to include custom filters for this value now we're looking to simplify.